### PR TITLE
Ironic fence agent

### PR DESCRIPTION
**Pull request by mbaldessari**
Introduce a fence agent that plugs into OpenStack's Ironic (Bare Metal
as a service) service.

Testing recap:
1) With ironic services up and running
1.1) Status
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o status -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Status: ON

1.2) Power off
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o off -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Success: Powered OFF  // Machine confirmed to be off

1.3) Power on
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o on -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Success: Powered ON

1.4) Power on (when already powered on)
fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o on -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Success: Already ON

2) With ironic services down/in error
2.1) Status
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o status -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Failed: Unable to obtain correct plug status or plug is not available

2.2) Power off
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o off -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Failed: Unable to obtain correct plug status or plug is not available

2.3) Power on
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o on -n 2b40a642-3d7b-4a43-b6a2-57a3967ebb1c
Failed: Unable to obtain correct plug status or plug is not available

3) Non existing VM
$ fence_ironic -a $OS_AUTH_URL -T admin -l admin -p $OS_PASSWORD -o on -n foo
Failed: Unable to obtain correct plug status or plug is not available

4) Verified that password is never logged even when invoked with '-v'

### CODE REVIEW

**Comment by mbaldessari** [comment]
Did not notice there is a make check ;) Let me fix that up and update this

File: fence/agents/ironic/fence_ironic.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    if "--tenant-name" in options:
+        env["OS_TENANT_NAME"] = options["--tenant-name"]
+
+    #  --auth-url -k
+    if "--auth-url" in options:
+        env["OS_AUTH_URL"] = options["--auth-url"]
+
+    # --action / -o
+    if action == "status":
+        cmd += " show %s -c power_state --format value" % (get_name_or_uuid(options))
```
Comment (jnpkrn) [inline fence/agents/ironic/fence_ironic.py]: AFAIK, `str.format` is preferred to `%` operator.

File: fence/agents/ironic/fence_ironic.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            pass
+        if "UUID" in uuid:
+            continue # skip line header
+        match = re.search('power[\\s]*([a-zA-Z]{2,3})', state)
+        status = match.group(1) if match else None
+        nodes[uuid] = (name, status)
+
+    return nodes
+
+def ironic_run_command(options, action, timeout=None):
```
Comment (marxsk) [inline fence/agents/ironic/fence_ironic.py]: Would it make sense to re-use/generalize
https://github.com/ClusterLabs/fence-agents/blob/master/fence/agents/amt/fence_amt.py#L39 
if yes then we can put it into fencing library.

File: fence/agents/ironic/fence_ironic.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    elif action == "list":
+        cmd += " list -c UUID -c Name -c 'Power State' --format csv --quote minimal"
+
+
+    logging.debug("cmd -> %s" % cmd)
+    return run_command(options, cmd, timeout, env)
+
+def define_new_opts():
+    all_opt["auth-url"] = {
+        "getopt" : "k:",
```
Comment (marxsk) [inline fence/agents/ironic/fence_ironic.py]: Please, do not introduce short options as we already have too many of them. That's the problem on our side.

File: fence/agents/ironic/fence_ironic.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        "longopt" : "tenant-name",
+        "help" : "-t, --tenant-name=[tenant]      Tenantname",
+        "required" : "0",
+        "shortdesc" : "Keystone Admin Tenant",
+        "default": "admin",
+        "order": 1
+    }
+    all_opt["openstackpath"] = {
+        "getopt" : ":",
+        "longopt" : "openstackpath",
```
Comment (marxsk) [inline fence/agents/ironic/fence_ironic.py]: Please, rename to openstack-path to follow our non-written conventions.

File: fence/agents/ironic/fence_ironic.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    env = os.environ.copy()
+    # --username / -l
+    if "--username" in options and len(options["--username"]) != 0:
+        env["OS_USERNAME"] = options["--username"]
+
+    # --password / -p
+    if "--password" in options:
+        x = quote(options["--password"])
+        x = x[:-1] if x.endswith("'") else x
+        x = x[1:] if x.startswith("'") else x
```
Comment (mbaldessari) [inline fence/agents/ironic/fence_ironic.py]: I kind of mimicked fence_amt here, but yeah might make sense to remove it.

File: fence/agents/ironic/fence_ironic.py (hunk 2 above)
Comment (mbaldessari) [inline fence/agents/ironic/fence_ironic.py]: Not sure it actually breaks if you have an apostrophe as a password, so maybe I should drop this anyway

File: fence/agents/ironic/fence_ironic.py (hunk 3 above)
Comment (mbaldessari) [inline fence/agents/ironic/fence_ironic.py]: Ack, thanks

File: fence/agents/ironic/fence_ironic.py (hunk 4 above)
Comment (mbaldessari) [inline fence/agents/ironic/fence_ironic.py]: Ack, makes sense

File: fence/agents/ironic/fence_ironic.py (hunk 2 above)
Comment (marxsk) [inline fence/agents/ironic/fence_ironic.py]: ok

**Pull request merged**